### DISCUSSION TITLE
The hast specification and implementation (hastscript) of Doctype (and Root) are different

### DISCUSSION THREAD

**Discussion by takahashim** [syntax-tree/unist#58] [Q&A] [answered]
Hi,

If we run a script like the following:

```js
import {h} from 'hastscript'
console.log(h('doctype'))
```

The result in console is as follows:

```js
{ type: 'element', tagName: 'doctype', properties: {}, children: [] }
```

However, in the hast specification, Doctype is a Doctype object that inherits from Node, not an Element or an object that inherits from Element.
https://github.com/syntax-tree/hast#doctype

Therefore, the specification seem to expect the following:

```js
{ type: 'doctype' }
```

Also the same thing seems to be happening for Root.
https://github.com/syntax-tree/hast#root

IMO there is a problem with either the hast spec or the hastscript. Which is correct?

**Comment by ChristianMurphy** [comment, marked as answer]
Both are correct.
The definition of doctype in hast is correct.
And hastscript generates elements exclusively.

If you want to add non-element hast type, like doctype, use https://github.com/syntax-tree/unist-builder

```js
import { u } from 'unist-builder';

console.log(u('doctype'));
```

**Reply by takahashim** [reply]
Thanks for your answer!
So, hastscript is used to generate HTML elements, and when generating HTML documents including doctype, unist-builder is also used together, right?

**Reply by ChristianMurphy** [reply]
Right, unist build and hastscript can be mixed and matched, make generating elements easy, and also add things generally not supported by hyperscript like doctype.

**Reply by takahashim** [reply]
Thank you!  I understand very well.
